Document InaccessibleObjectException/IllegalAccessException on JDK 16+ with solution (#965)
mockk/mockk: doc/md/jdk16-access-exceptions.md

# JDK 16+ access exceptions

### Problem symptoms 

> On JDK 16 and above you may encounter InaccessibleObjectException or IllegalAccessException in following known cases:

Reviewer: ```suggestion
On JDK 16 and above you may encounter `InaccessibleObjectException` or `IllegalAccessException` in following known cases:
```

> On JDK 16 and above you may encounter `InaccessibleObjectException` or `IllegalAccessException` in following known cases:

Reviewer 2: "in the following known cases"

> 1. _Some_ usages of `mockkStatic` on java standard library classes, e.g. 

Reviewer: ```suggestion
1. _Some_ usages of `mockkStatic` on Java standard library classes, e.g. 
```

   ```kotlin
   mockkStatic(Instant::class)
   every { Instant.now() } returns Instant.parse("2022-09-15T09:31:26.919Z")
   ```
   ```
   java.time.format.DateTimeParseException: Text '2022-09-15T09:31:26.919Z' could not be parsed:
     Unable to make private static java.time.Instant java.time.Instant.create(long,int) accessible: 
     module java.base does not "opens java.time" to unnamed module @4501b7af
   ```

> 2. Spying on java standard library classes, e.g.

Reviewer: ```suggestion
2. Spying on Java standard library classes, e.g.
```

   ```kotlin
   val socket = Socket()
   val spy = spyk<Socket>(socket)
   ```
    Note: this includes creating `@SpyKBean` on spring data repository instances, because they are instances of `java.lang.reflect.Proxy` in fact.
   ```
   Error creating bean with name 'featureRepository': Post-processing of FactoryBean's singleton object failed;
    nested exception is java.lang.IllegalAccessException: class io.mockk.impl.InternalPlatform 
    cannot access a member of class java.lang.reflect.Proxy (in module java.base) with modifiers "protected"
   ```

### Problem cause

> JDK 16 enforced strong encapsulation of standard modules. In practice this means you (and your libraries) can no more get away with call like `privateMember.setAccessible(true)` to any of JDK classes.

Reviewer: ```suggestion
JDK 16 enforces strong encapsulation of standard modules. In practice this means libraries are forbidden from overriding accessibility modifiers in JDK classes (for example,  `privateMember.setAccessible(true)`).
```

Author: > In practice this means libraries are forbidden

AFAIK not only libraries. Do you want to make accent on mockk use-case?

Reviewer: Good point, I've edited it. 

I think the project name should be styled as: MockK

https://blogs.oracle.com/javamagazine/post/a-peek-into-java-17-continuing-the-drive-to-encapsulate-the-java-runtime-internals

### Solution / workaround

> Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need  

Reviewer: ```suggestion
Add JVM argument `--add-opens java.base/java.time=ALL-UNNAMED` _for each package_ you need to mock.
```

Author: Agree.
And to be more general here, we also can replace `java.time` with `package.name`. And there are specific examples later in the file

Reviewer: Looking at the [docs here](https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-2F61F3A9-0979-46A4-8B49-325BA0EE8B66), would the flag `--illegal-access=warn` work better?

Author: They say `--illegal-access` was removed in JDK 17

[Article](https://blogs.oracle.com/javamagazine/post/a-peek-into-java-17-continuing-the-drive-to-encapsulate-the-java-runtime-internals), section "What’s changing in Java 17?"

Because of this I did not mention this flag. I think that JDK 17 is going to be much more popular, and current users of JDK 16 should not rely on `--illegal-access` option, in order not to face access problems again in future.

Reviewer: Ah clear, thanks for explaining.

Description: https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html

> For gradle users:

Reviewer: ```suggestion
For Gradle users:
```

```groovy
tasks.test {
    jvmArgs("--add-opens", "java.base/java.time=ALL-UNNAMED")
}

> ```

Author: I am not very familiar with gradle, and copied this from one of the issues. May need help from experienced gradle user to verify that this snippet looks ok

Reviewer: Yeah it looks okay. It will only change the 'test' class though. In .kts, this is the best way to change all test tasks:

```kts
// build.gradle.kts

tasks.withType<Test>().configureEach {
  jvmArgs(...)
}
```

> For maven users:

Reviewer: ```suggestion
For Maven users:
```
